### **Prediction** — held-out weighted-crime-rate model

Target: **`log(weighted rate)`** at the block-group level (absolute, no demeaning). We predict a BG's
crime from **BG-level predictors** plus, in the top rung, the **agency-wide lagged crime total** — one
value that is the *same for every BG in a city*, so it carries the city-level (between-city) share
while the BG predictors explain the within-city variation.

Every configuration is scored on the **two questions `04_bg_comparison` used to indict the incumbent**:

| axis | question | metrics |
| --- | --- | --- |
| **LEVEL** (cross-city) | are whole cities placed at the right height? | `level_r`, `r2_oos`, `mae` |
| **RANK** (within-city) | inside a city, are the dangerous blocks flagged? | `recall@top10/30`, `called_safe`, `withinR2` |

- **City sets:** `wtotal` / 5 full-coverage cities; `wprop` / all property-covered cities.
- **Splits:** stratified **80/20** (interpolation — optimistic) and **LOCO** (leave-one-city-out,
  extrapolation — honest). The **gap** between them is the value of having seen a city before.
- **Predictor ladder:** `ours` → `ours+approved` (de-duplicated incumbent features) →
  `ours+approved+agency` (adds the agency lagged total). No race / protected predictors.
- **Models:** `RidgeCV` then `LightGBM`, same ladder.

In [1]:
%load_ext autoreload
%autoreload 2
%matplotlib inline

import contextlib
import io
import sys
import warnings
from pathlib import Path

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

_here = Path.cwd().resolve()
_root = next(p for p in [_here, *_here.parents] if (p / "src").is_dir())
sys.path.insert(0, str(_root / "src"))

from crime_blockgroup_mapping.constants import CITIES
from regression_modelling.constants import AGENCY_ANCHOR_COL
from regression_modelling.models.cv import load_pooled_table, loco_metrics, within_city_recall
from regression_modelling.models.experiments import (
    ExperimentSpec, run_spec, between_city_level_r, resolve_predictors,
)

pd.set_option("display.width", 240)
pd.set_option("display.max_columns", 40)
warnings.filterwarnings("ignore", category=RuntimeWarning)
warnings.filterwarnings("ignore", message="An input array is constant")

# 5 full-coverage cities carry weighted-total (wtotal); the rest add weighted-property (wprop).
full_cities = [c for c, cfg in CITIES.items() if not cfg.property_only]
all_cities = list(CITIES.keys())

# Pooled block-group tables: one row per BG with predictors, observed rates, and the agency total.
pool5 = load_pooled_table(full_cities)
pool10 = load_pooled_table(all_cities)

# Drop cities whose crime counts are entirely zero. Columbus's upstream bg_crime extract produced
# all-zero counts (BG geometry present, crime never joined) -> a constant rate that makes
# correlation / ranking metrics undefined. Remove it rather than let it poison the LEVEL metrics.
_degenerate = [c for c, g in pool10.groupby("city") if float(g["wprop_rate"].abs().sum()) == 0]
if _degenerate:
    print("dropping all-zero-crime cities (upstream data gap):", _degenerate)
    pool10 = pool10[~pool10["city"].isin(_degenerate)].copy()
    all_cities = [c for c in all_cities if c not in _degenerate]

print("pool5:", pool5.shape, "| pool10:", pool10.shape, "| wprop cities:", len(all_cities))

  houston            1629 BGs
  chicago            2164 BGs
  atlanta             426 BGs
  kansas_city         473 BGs
  detroit             622 BGs
  pooled             5314 BGs across 5 cities
  dropped 22 zero/NaN-pop BGs -> 5292 remain (filtered geoid set for fit + Moran's I + bias join)
  dropped 8 BGs below daytime_pop 100 (small-denominator rate artifacts) -> 5284 remain
  houston            1629 BGs
  chicago            2164 BGs
  atlanta             426 BGs
  sacramento          366 BGs
  san_francisco       679 BGs
  pittsburgh          317 BGs
  kansas_city         473 BGs
  columbus            637 BGs
  jacksonville        550 BGs


  detroit             622 BGs
  pooled             7863 BGs across 10 cities
  dropped 32 zero/NaN-pop BGs -> 7831 remain (filtered geoid set for fit + Moran's I + bias join)
  dropped 12 BGs below daytime_pop 100 (small-denominator rate artifacts) -> 7819 remain
dropping all-zero-crime cities (upstream data gap): ['columbus']
pool5: (5284, 108) | pool10: (7182, 108) | wprop cities: 9


### Benchmark & metric glossary

The `existing (04)*` row is the deployed **agency-scale** model (wtotal, 5 cities), scored
**in-sample**. Our rows are honest **LOCO** (city fully held out) — so *beating the incumbent
under LOCO is the strong result*.

- **`level_r`** — Pearson of city-mean predicted vs observed rate (cross-city LEVEL).
- **`r2_oos` / `mae`** — pooled out-of-sample fit on the absolute log-rate.
- **`recall@top10/30`** — of a city's truly-dangerous blocks (top decile by observed rate),
  the share the model also ranks in its own within-city top 10% / 30% (higher = better).
- **`called_safe`** — share of dangerous blocks buried below the model's within-city median
  (the costly miss; lower = better).
- **`withinR2`** — mean per-city (corr of predicted vs observed log-rate)² (within-city fit).

In [2]:
# Existing agency-scale model (04_bg_comparison), wtotal / 5 cities, scored IN-SAMPLE.
# Fixed reference row so the new honest-LOCO numbers sit beside the incumbent.
EXISTING_04 = {
    "level_r": 0.96,                                                    # 04 cell 7 (Pearson of city means)
    "recall@top10": 0.29, "recall@top30": 0.65, "called_safe": 0.16,   # 04 cell 18 POOLED
    "withinR2": 0.32,                                                   # 04 cell 11 (mean per-city R2)
}

_POOL = {"wtotal": pool5, "wprop": pool10}
_RUNS = {}


def get_run(category, pset, split, model="ridge"):
    """Return a cached absolute log-rate run for one (predictor set, split, model).

    Runs are memoised in ``_RUNS`` so each configuration is fit once, and fitter stdout is
    suppressed to keep tables clean. ``pset`` is 'ours' | 'ours+approved' |
    'ours+approved+agency'; ``split`` is 'loco' | '80_20'; ``model`` is 'ridge' | 'gbm'.
    """
    key = (category, pset, split, model)
    if key not in _RUNS:
        spec = ExperimentSpec(pset, "absolute", split, model)
        with contextlib.redirect_stdout(io.StringIO()):
            _RUNS[key] = run_spec(spec, _POOL[category], category)
    return _RUNS[key]


def within_city_r2(scored, category):
    """Mean per-city within-city R2 = (Pearson of y_pred vs log1p(rate))^2.

    Affine-invariant, so it needs no city moments — the honest 'within-city variance
    explained', comparable to 04's mean per-city R2. Returns the mean across cities.
    """
    rate = f"{category}_rate"
    r2s = []
    # One held-out city at a time: correlate predicted vs observed log-rate within that city.
    for _, g in scored.groupby("holdout_city"):
        if len(g) <= 2:
            continue
        y = np.log1p(g[rate].astype(float))
        p = g["y_pred"].astype(float)
        r = np.corrcoef(y, p)[0, 1]
        if np.isnan(r):          # constant-rate city (e.g. Columbus wprop) -> undefined, skip
            continue
        r2s.append(r * r)
    return round(float(np.mean(r2s)), 3) if r2s else np.nan


def combined_table(category, sets, model="ridge", existing=None):
    """Level+rank scorecard for one model across predictor sets, LOCO and 80/20.

    LEVEL columns (whole-city placement): level_r, r2_oos, mae.
    RANK columns (dangerous-block flagging within a city): recall@top10, recall@top30,
    called_safe, withinR2. An optional ``existing`` dict prepends the in-sample incumbent.
    """
    rows = []
    if existing is not None:
        rows.append({"predictors": "existing (04)*", "split": "in-sample",
                     "level_r": existing["level_r"], "r2_oos": np.nan, "mae": np.nan,
                     "recall@top10": existing["recall@top10"], "recall@top30": existing["recall@top30"],
                     "called_safe": existing["called_safe"], "withinR2": existing["withinR2"]})
    # One block per predictor set; within each, the honest LOCO row then the optimistic 80/20 row.
    for name in sets:
        for split in ["loco", "80_20"]:
            run = get_run(category, name, split, model)
            m = loco_metrics(run).loc["POOLED"]
            rec = within_city_recall(run, category).loc["POOLED"]
            rows.append({"predictors": name, "split": split.replace("_", "/"),
                         "level_r": between_city_level_r(run),
                         "r2_oos": round(float(m["r2_oos"]), 3), "mae": round(float(m["mae"]), 2),
                         "recall@top10": rec["recall@top10"], "recall@top30": rec["recall@top30"],
                         "called_safe": rec["called_safe"],
                         "withinR2": within_city_r2(run["scored"], category)})
    return pd.DataFrame(rows).set_index(["predictors", "split"])


def level_by_city(category, pset, model="ridge", split="loco"):
    """Per-city LEVEL + within-city fit for one config — decomposes the single `level_r`.

    For each held-out city returns its observed vs predicted city-MEAN log-rate (the level
    the agency total otherwise carries) and the within-city correlation (ranking quality). Makes
    visible whether a model's level skill is uniform or carried by 1-2 cities, and exposes
    the shrinkage-vs-calibration contrast (predicted spread vs observed spread).
    """
    run = get_run(category, pset, split, model)
    rate = f"{category}_rate"
    g = run["scored"].copy()
    g["_obs"] = np.log1p(g[rate].astype(float))
    rows = []
    # One row per held-out city: its mean level (obs vs pred) and within-city rank correlation.
    for city, d in g.groupby("holdout_city"):
        r = np.corrcoef(d["_obs"], d["y_pred"])[0, 1] if len(d) > 2 else np.nan
        rows.append({"city": city, "n": len(d),
                     "obs_mean": round(float(d["_obs"].mean()), 3),
                     "pred_mean": round(float(d["y_pred"].mean()), 3),
                     "level_err": round(float(d["y_pred"].mean() - d["_obs"].mean()), 3),
                     "within_r": round(float(r), 3)})
    out = pd.DataFrame(rows).set_index("city")
    # Between-city level_r = correlation of the per-city obs vs pred means (the 5%/10% axis).
    out.attrs["level_r"] = round(float(np.corrcoef(out["obs_mean"], out["pred_mean"])[0, 1]), 3)
    return out

### Predictor sets

Exactly what each rung feeds the model. The ladder is additive: `ours+approved` adds the approved,
de-duplicated incumbent features to `ours`; `ours+approved+agency` adds **one** column,
`agency_lag_<target>_log` — last year's agency-wide crime total, identical across every BG in a city
(so it carries the city-level variance; the BG predictors carry the rest).

In [3]:
# Print the resolved predictor columns for each rung (wtotal pool), showing what each rung adds.
_seen = []
for pset in ["ours", "ours+approved", "ours+approved+agency"]:
    cols = resolve_predictors(ExperimentSpec(pset, "absolute", "loco", "ridge"), "wtotal", pool5)
    added = [c for c in cols if c not in _seen]      # only the columns this rung introduces
    print(f"{pset}  (total {len(cols)}, +{len(added)} new):")
    for c in added:
        print("   +", c)
    _seen = cols
    print()

ours  (total 23, +23 new):
   + moved1yr_pct
   + own_pct
   + lap_pct
   + city_centers_dist
   + pop_est_5mile_log
   + pop_ch_1mile
   + vacant_pct_log
   + clip_liens_pct_log
   + clip_foreclosure_pct_log
   + unq_convenience_stores_clips
   + unq_gas_stations_clips
   + unq_liquor_stores_clips
   + vacant_pct_lag6_log
   + clip_liens_pct_lag6_log
   + clip_foreclosure_pct_lag6_log
   + transit_has_transit
   + transit_service_intensity_logc
   + transit_nearest_stop_m_log
   + transit_overnight_stop_share
   + roof_condition_avg
   + roof_debris_pct_avg
   + hardscapes_pct_avg
   + roof_missing_material_pct

ours+approved  (total 28, +5 new):
   + in_household_pct
   + det_pct
   + div_encentral
   + div_midatlantic
   + div_southatlantic

ours+approved+agency  (total 29, +1 new):
   + agency_lag_wtotal_log



### Step 1 — RidgeCV across the predictor ladder

Two questions at once:
1. **Do the approved + de-duplicated incumbent features earn their place?** (`ours` → `ours+approved`)
2. **Does adding the agency lagged total explain the city-level variance?** (`ours+approved` → `+agency`)

`agency_lag_wtotal_log` is last year's observed agency-wide crime — a real per-city number (not learned
from the other cities), constant across a city's BGs. Watch `level_r` and `r2_oos` on the **LOCO** rows
as it is added (n = 5 cities — a single between-city predictor is powerful but fragile; read as directional).

In [4]:
SETS = ["ours", "ours+approved", "ours+approved+agency"]
print("CROSS-CITY · absolute log(weighted rate) · wtotal / 5 full-coverage cities · RidgeCV")
print(combined_table("wtotal", SETS, model="ridge", existing=EXISTING_04).to_string())
print("\n* existing (04) is IN-SAMPLE (level_r = Pearson of city means; recall/withinR2 from 04).")
print("  ours / +approved / +agency rows are honest LOCO and optimistic 80/20.")

CROSS-CITY · absolute log(weighted rate) · wtotal / 5 full-coverage cities · RidgeCV


                                level_r  r2_oos   mae  recall@top10  recall@top30  called_safe  withinR2
predictors           split                                                                              
existing (04)*       in-sample    0.960     NaN   NaN         0.290         0.650        0.160     0.320
ours                 loco         0.139   0.245  0.72         0.421         0.788        0.077     0.422
                     80/20        0.902   0.502  0.60         0.304         0.775        0.049     0.447
ours+approved        loco        -0.023   0.232  0.72         0.409         0.782        0.079     0.428
                     80/20        0.957   0.514  0.59         0.314         0.775        0.069     0.448
ours+approved+agency loco         0.194   0.194  0.75         0.411         0.784        0.085     0.426
                     80/20        0.963   0.515  0.59         0.324         0.765        0.069     0.449

* existing (04) is IN-SAMPLE (level_r = Pearson of cit

**Read — RidgeCV.**

- **RANK (the product) — we beat the incumbent decisively, under honest LOCO.** Across the ladder,
  `recall@top10` ≈ **0.41** vs existing **0.29**, `recall@top30` ≈ **0.78** vs **0.65**, `called_safe`
  ≈ **0.08** vs **0.16** (half the costly misses), `withinR2` ≈ **0.43** vs **0.32**. This is the
  block-level discrimination `04` showed the agency-scale model lacks.
- **The approved + de-duped features barely move RANK** (recall/withinR2 flat) and actually **hurt
  cross-city LEVEL under LOCO** (`level_r` 0.14 → −0.02): a held-out city's level can't be inferred
  from the others. They *do* help 80/20 LEVEL (0.90 → 0.96). Net: harmless to the product, not
  earning their place on extrapolated LEVEL.
- **Adding the agency lagged total barely moves LOCO LEVEL under RidgeCV** (`level_r` 0.14 → 0.19;
  `r2_oos` actually *drops* 0.245 → 0.194). The column is in the model, but **Ridge shrinks its
  coefficient** toward 0, so with only 5 cities the level correction is damped. (LightGBM uses it far
  more — Step 2.)
- **LEVEL is almost entirely "have we seen this city":** the 80/20→LOCO `level_r` gap is huge
  (≈ 0.90 → 0.14) — the city-level share needs the agency total (used properly) or a seen city, not
  the BG predictors.


### Step 2 — LightGBM, same ladder

Same three predictor sets, trees instead of a linear model. LightGBM is level-/scale-invariant
(predictors enter raw), so this isolates **model form**: does a non-linear learner beat RidgeCV on
either LEVEL or RANK, and does it use the agency lagged total to place unseen cities?

In [5]:
print("CROSS-CITY · absolute log(weighted rate) · wtotal / 5 cities · LightGBM")
print(combined_table("wtotal", SETS, model="gbm", existing=EXISTING_04).to_string())
print("\n* same predictor ladder as Step 1; LightGBM replaces RidgeCV. Compare LOCO rows head-to-head.")

CROSS-CITY · absolute log(weighted rate) · wtotal / 5 cities · LightGBM


                                level_r  r2_oos   mae  recall@top10  recall@top30  called_safe  withinR2
predictors           split                                                                              
existing (04)*       in-sample    0.960     NaN   NaN         0.290         0.650        0.160     0.320
ours                 loco         0.592   0.353  0.70         0.363         0.758        0.109     0.380
                     80/20        0.952   0.516  0.60         0.333         0.735        0.039     0.461
ours+approved        loco         0.569   0.379  0.68         0.393         0.772        0.095     0.404
                     80/20        0.948   0.521  0.60         0.392         0.745        0.049     0.472
ours+approved+agency loco         0.691   0.375  0.69         0.397         0.758        0.101     0.398
                     80/20        0.966   0.529  0.60         0.363         0.745        0.069     0.475

* same predictor ladder as Step 1; LightGBM replaces R

**Read — RidgeCV vs LightGBM.**

- **Model form matters for LEVEL.** LightGBM lifts LOCO `level_r` from RidgeCV's ≈ 0.14 to **0.57–0.69**
  (`r2_oos` 0.19–0.25 → **0.35–0.38**). Trees actually *use* the agency lagged total (and non-linear
  predictor interactions) to place cities — Ridge underuses it via shrinkage. The agency rung pushes
  GBM's LOCO `level_r` to its best **0.69**.
- **But RidgeCV wins RANK — the product.** GBM's LOCO `recall@top10` ≈ 0.36–0.40 (< Ridge ≈ 0.41),
  `withinR2` ≈ 0.38–0.40 (< Ridge ≈ 0.43), `called_safe` ≈ 0.10 (> Ridge ≈ 0.08). The linear model
  ranks blocks inside a city slightly better and more stably (smaller 80/20→LOCO gap).
- **Caveat:** n = 5 cities — the LOCO LEVEL numbers turn on 1–2 cities (see Step 4). Step 3 re-runs the
  whole thing on the wider property target to see which results survive.
- **Both models beat the incumbent on RANK under LOCO.**


### Step 3 — wider evidence: the property target, all covered cities

The wtotal verdict rests on 5 cities, so `level_r` there is a 5-point correlation (fragile). The
weighted-**property** target covers the wider POC set, turning `level_r` into a less brittle read.
Same predictor ladder, same two models. No `existing (04)` row — `04` benchmarked wtotal/5 only.
(Different target/scorecard from wtotal: compare *within* this table, never cross-target R²/MAE.)

In [6]:
print(f"CROSS-CITY · absolute log(weighted PROPERTY rate) · wprop / {len(all_cities)} cities · RidgeCV")
print(combined_table("wprop", SETS, model="ridge").to_string())

CROSS-CITY · absolute log(weighted PROPERTY rate) · wprop / 9 cities · RidgeCV


                            level_r  r2_oos   mae  recall@top10  recall@top30  called_safe  withinR2
predictors           split                                                                          
ours                 loco     0.550   0.327  0.59         0.339         0.720        0.119     0.325
                     80/20    0.848   0.385  0.57         0.369         0.702        0.128     0.333
ours+approved        loco     0.797   0.358  0.58         0.352         0.728        0.112     0.337
                     80/20    0.957   0.417  0.56         0.411         0.709        0.142     0.355
ours+approved+agency loco     0.964   0.380  0.57         0.357         0.725        0.113     0.336
                     80/20    0.987   0.422  0.55         0.390         0.709        0.142     0.354


In [7]:
print(f"CROSS-CITY · absolute log(weighted PROPERTY rate) · wprop / {len(all_cities)} cities · LightGBM")
print(combined_table("wprop", SETS, model="gbm").to_string())

CROSS-CITY · absolute log(weighted PROPERTY rate) · wprop / 9 cities · LightGBM


                            level_r  r2_oos   mae  recall@top10  recall@top30  called_safe  withinR2
predictors           split                                                                          
ours                 loco     0.291   0.270  0.63         0.342         0.701        0.117     0.349
                     80/20    0.965   0.460  0.54         0.348         0.723        0.135     0.402
ours+approved        loco     0.659   0.360  0.59         0.349         0.709        0.114     0.365
                     80/20    0.990   0.483  0.53         0.397         0.738        0.099     0.428
ours+approved+agency loco     0.789   0.359  0.59         0.367         0.709        0.123     0.358
                     80/20    0.991   0.490  0.52         0.390         0.730        0.106     0.432


**Read — wider property target (9 cities, Columbus dropped).**

With the degenerate city gone, the property target tells a **cleaner, more decisive story than wtotal/5**:

- **The agency lagged total is a decisive between-city LEVEL predictor — and RidgeCV uses it fully here.**
  Ridge LOCO `level_r` climbs monotonically **0.55 → 0.80 → 0.96** across `ours → +approved → +agency`
  (`r2_oos` 0.33 → 0.36 → 0.38). This is exactly the intended design: BG predictors explain within-city
  variation, the one agency column carries the city-level share. The earlier "Ridge shrinks agency away"
  read was a 5-city/Columbus artifact — on the wider target the coefficient is not damped.
- **On this wider target RidgeCV actually beats LightGBM on LEVEL.** GBM LOCO `level_r` is
  0.29 → 0.66 → **0.79** — it uses the agency total but tops out below Ridge's **0.96**. So the wtotal/5
  "GBM wins LEVEL" (0.69 vs 0.19) did **not** replicate; it was driven by Atlanta + only-5-cities.
- **RANK is a near-tie, both models solid.** LOCO `recall@top10` ≈ 0.34–0.37, `recall@top30` ≈ 0.71–0.73,
  `withinR2` ≈ 0.34 (Ridge) vs ≈ 0.36 (GBM). No inversions this time — Columbus was causing the earlier
  Ridge SF/Detroit collapse.
- **Verdict shift:** on the less-brittle target, the linear model + agency predictor gives the best
  calibrated LEVEL *and* competitive RANK. GBM's earlier LEVEL edge does not survive Columbus removal.


### Step 4 — per-city breakdown: is `level_r` uniform or carried by 1-2 cities?

`level_r` is one correlation over 5 (wtotal) or 10 (wprop) city-mean points, so a single city can
swing it. `level_by_city` opens it up per held-out city:
- **`level_err`** = predicted mean − observed mean (LOCO). Near 0 = the model places the city at the
  right height; a large-magnitude column that flips sign across cities is what makes `level_r` noisy.
- **`within_r`** = within-city correlation (ranking quality), the offset-free product axis.

Watch two things: (1) whether Ridge's `level_err` is *systematically* compressed toward 0-spread
(shrink-to-mean) while GBM's tracks the observed spread (calibration); (2) whether GBM's `level_r`
edge is uniform or rides on a couple of cities.

In [8]:
# Compare the two models' per-city level + within-city fit on the top rung (ours+approved+agency).
CFG = "ours+approved+agency"
for cat in ["wtotal", "wprop"]:
    # One target at a time; print Ridge then GBM so the level_err / within_r columns sit side by side.
    for model in ["ridge", "gbm"]:
        tbl = level_by_city(cat, CFG, model=model)
        print(f"== {cat} · {CFG} · {model} · LOCO (between-city level_r = {tbl.attrs['level_r']}) ==")
        print(tbl.to_string())
        print()

== wtotal · ours+approved+agency · ridge · LOCO (between-city level_r = 0.194) ==
                n  obs_mean  pred_mean  level_err  within_r
city                                                       
atlanta       369     3.414      5.279      1.865     0.715
chicago      2064     3.443      3.100     -0.344     0.690
detroit       600     4.201      4.548      0.347     0.457
houston      1536     3.446      3.594      0.148     0.593
kansas_city   445     3.742      3.566     -0.176     0.763

== wtotal · ours+approved+agency · gbm · LOCO (between-city level_r = 0.691) ==
                n  obs_mean  pred_mean  level_err  within_r
city                                                       
atlanta       369     3.414      3.987      0.573     0.661
chicago      2064     3.443      3.593      0.150     0.666
detroit       600     4.201      4.488      0.287     0.471
houston      1536     3.446      3.283     -0.163     0.576
kansas_city   445     3.742      3.441     -0.302     0.7

**Read — per-city breakdown (why `level_r` moves).** `level_r` is a correlation over a handful of
city means, so one miscalibrated city dominates it.

- **wtotal / 5 cities — Ridge is sunk by a single city.** Atlanta `level_err` **+1.865** (predicts 5.28
  vs observed 3.41) drags Ridge's `level_r` to **0.194** while the other four sit within ±0.35. GBM has
  no blow-up (worst |err| 0.573) → `level_r` **0.691**. This lone outlier — not a general skill gap — is
  the whole wtotal LEVEL story. Do **not** tune toward `level_r` on n = 5.
- **wprop / 9 cities — with the agency total, Ridge is well-calibrated everywhere.** Every city's
  `level_err` is within ±0.17 (max Detroit +0.136) → `level_r` **0.964**. GBM instead over-predicts
  Atlanta (**+0.535**) and under-predicts Kansas City (−0.316) → `level_r` **0.789**. The linear model's
  even calibration across cities is what wins LEVEL on the wider target.
- **`within_r` is stable and healthy for both** (≈ 0.37–0.77 per city, no negatives) — the RANK product
  holds up city-by-city, independent of the LEVEL wobble.
- **Takeaway:** LEVEL rides on a few city means and on the agency predictor being used at full strength;
  RANK is broad-based and robust. Judge extrapolation on the 9-city property evidence, not wtotal/5.
